Fix the failure memory order argument to compareExchangeStrong

LumioseSil opened the pull request:
#### d2ba4a07e209c5cb422ff5f5c50802e1f0582438
<pre>
Fix the failure memory order argument to compareExchangeStrong
<a href="https://bugs.webkit.org/show_bug.cgi?id=266649">https://bugs.webkit.org/show_bug.cgi?id=266649</a>

Reviewed by Kimmo Kinnunen.

The failure memory order cannot be release or acq_rel.
Clang since llvm/llvm-project@fed5644 diagnoses an invalid argument.
Furthermore, the current way it is written is undefined behavior anyway.

From <a href="https://en.cppreference.com/w/cpp/atomic/atomic/compare_exchange">https://en.cppreference.com/w/cpp/atomic/atomic/compare_exchange</a>:
&quot;If failure is stronger than success or(until C++17) is one of
std::memory_order_release and std::memory_order_acq_rel, the behavior is
undefined.&quot;

* Source/WebKit/Platform/IPC/StreamClientConnectionBuffer.h:
  (IPC::StreamClientConnectionBuffer::tryAcquire): Use
  memory_order_acquire as fallback.
* Source/WebKit/Platform/IPC/StreamServerConnectionBuffer.h:
  (IPC::StreamServerConnectionBuffer::tryAcquire): Ditto.

Canonical link: <a href="https://commits.webkit.org/272461@main">https://commits.webkit.org/272461@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/83b9eb43ef0598bc13c9beef156358b9459925e6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/31709 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/10392 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/33434 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/34202 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/28714 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/12749 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/51/builds/7640 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/34202 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/32062 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/12749 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/33434 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/34202 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/12749 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/33434 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/35547 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/12749 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/33434 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/35547 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/7813 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/51/builds/7640 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/35547 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/9468 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/33434 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7433 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/8492 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/8329 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas reviewed (commented):
This will need a review from someone who's familiar with `WTF::Atomic`.

I _think_ the change is good, but you should have explained better in your changelog that the status quo is undefined behavior. Can you update your commit message to reflect that?

From https://en.cppreference.com/w/cpp/atomic/atomic/compare_exchange

> If failure is stronger than success or(until C++17) is one of [std::memory_order_release](http://en.cppreference.com/w/cpp/atomic/memory_order) and [std::memory_order_acq_rel](http://en.cppreference.com/w/cpp/atomic/memory_order), the behavior is undefined.

LumioseSil committed:
Fix the failure memory order argument to compareExchangeStrong
https://bugs.webkit.org/show_bug.cgi?id=266649

Reviewed by Kimmo Kinnunen.

The failure memory order cannot be release or acq_rel.
Clang since llvm/llvm-project@fed5644 diagnoses an invalid argument.
Furthermore, the current way it is written is undefined behavior anyway.

From https://en.cppreference.com/w/cpp/atomic/atomic/compare_exchange:
"If failure is stronger than success or(until C++17) is one of
std::memory_order_release and std::memory_order_acq_rel, the behavior is
undefined."

* Source/WebKit/Platform/IPC/StreamClientConnectionBuffer.h:
  (IPC::StreamClientConnectionBuffer::tryAcquire): Use
  memory_order_acquire as fallback.
* Source/WebKit/Platform/IPC/StreamServerConnectionBuffer.h:
  (IPC::StreamServerConnectionBuffer::tryAcquire): Ditto.

Canonical link: https://commits.webkit.org/272461@main
